Core/src/Utilities/FpeMonitor.cpp

```diff
@@ -10,81 +10,255 @@
+  std::cout << std::endl;
+
+  FpeMonitor &fpe = *stack().top();
+  fpe.m_result.m_counts.at(si->si_code)++;
+
+  std::size_t maxDepth = static_cast<std::size_t>(-1);
+  maxDepth = 16;
+
+  // collect stack trace skipping 2 frames, which should be the signal handler
+  // and the calling facility. This might be platform specific, not sure
+  if (fpe.m_result.m_stracktraces.size() < fpe.stackLimit()) {
+    fpe.m_result.m_stracktraces.push_back(
+        {1, static_cast<FpeType>(si->si_code), StackTrace(2, maxDepth)});
+  }
+
+  // @TODO: Disable this on non-x86_64
+  __uint16_t *cw = &((ucontext_t *)ctx)->uc_mcontext.fpregs->cwd;
+  *cw |= FPU_EXCEPTION_MASK;
+
+  __uint16_t *sw = &((ucontext_t *)ctx)->uc_mcontext.fpregs->swd;
+  *sw &= ~FPU_STATUS_FLAGS;
+
+  __uint32_t *mxcsr = &((ucontext_t *)ctx)->uc_mcontext.fpregs->mxcsr;
+  // *mxcsr |= SSE_EXCEPTION_MASK;  // disable all SSE exceptions
+  *mxcsr |= ((*mxcsr & SSE_STATUS_FLAGS) << 7);
+  *mxcsr &= ~SSE_STATUS_FLAGS;  // clear all pending SSE exceptions
 }
 
-void FpeMonitor::enable(int excepts) {
+void FpeMonitor::enable() {
+  // @TODO: Disable this on non-x86_64
 #if defined(__APPLE__)
   std::cerr << "FPE monitoring currently not supported on Apple" << std::endl;
-  (void)excepts;
-  (void)handle_fpe;
 #else
-  std::feclearexcept(FE_ALL_EXCEPT);
-  feenableexcept(excepts);
+  std::feclearexcept(m_excepts);
+  feenableexcept(m_excepts);
+
+  ensureSignalHandlerInstalled();
 
-  std::signal(SIGFPE, handle_fpe);
+  stack().push(this);
 #endif
 }
-void FpeMonitor::disable(int excepts) {
+
+void FpeMonitor::rearm() const {
+  feenableexcept(m_excepts);
+}
+
+void FpeMonitor::ensureSignalHandlerInstalled() {
+  auto &state = globalState();
+  if (state.isSignalHandlerInstalled) {
+    return;
+  }
+
+  std::lock_guard lock{state.mutex};
+
+  struct sigaction action;
+  action.sa_sigaction = &signalHandler;
+  action.sa_flags = SA_SIGINFO;
+  sigaction(SIGFPE, &action, nullptr);
+
+  state.isSignalHandlerInstalled = true;
+}
+
+void FpeMonitor::disable() {
+  // @TODO: Disable this on non-x86_64
 #if defined(__APPLE__)
   std::cerr << "FPE monitoring currently not supported on Apple" << std::endl;
-  (void)excepts;
 #else
-  fedisableexcept(excepts);
-  std::signal(SIGFPE, SIG_DFL);
+  std::feclearexcept(m_excepts);
+  // fedisableexcept(m_excepts);
+  stack().pop();
 #endif
+
+  // std::cout << std::bitset<32>(m_impl->m_encountered) << std::endl;
+}
+
+std::stack<FpeMonitor *> &FpeMonitor::stack() {
+  static thread_local std::stack<FpeMonitor *> monitors;
+  return monitors;
+}
+
+FpeMonitor::GlobalState &FpeMonitor::globalState() {
+  static GlobalState state{};
+  return state;
+}
+
+std::ostream &operator<<(std::ostream &os, FpeType type) {
+#define CASE(x)    \
+  case FpeType::x: \
+    os << #x;      \
+    break;
+
+  switch (type) {
+    CASE(INTDIV)
+    CASE(INTOVF)
+    CASE(FLTDIV)
+    CASE(FLTOVF)
+    CASE(FLTUND)
+    CASE(FLTRES)
+    CASE(FLTINV)
+    CASE(FLTSUB)
+  }
```

Comment:
does a `default` case make sense here?

Reply:
Should be exhaustive but could add it, yeah.